api/envoy/service/ext_proc/v3/external_processor.proto

```diff
@@ -362,14 +362,29 @@ message HeaderMutation {
   repeated string remove_headers = 2;
 }
 
-// Replace the entire message body chunk received in the corresponding
-// HttpBody message with this new body, or clear the body.
+// The body response message corresponding to BIDIRECTIONAL_STREAMED body mode.
+message StreamedBodyResponse {
+  // Envoy sends this body mutation out as it is.
+  bytes body = 1;
+
+  // Set this flag to true if this is the last chunk of body responses,
+  // and there is no trailers behind it.
+  bool end_of_stream = 2;
+}
+
+// This message specifies the body mutation Envoy will apply and send out.
 message BodyMutation {
+  // Sending this response for any body mode except BIDIRECTIONAL_STREAMED.
   oneof mutation {
     // The entire body to replace
     bytes body = 1;
 
     // Clear the corresponding body chunk
     bool clear_body = 2;
   }
+
+  // [#not-implemented-hide:]
+  // Sending this response for BIDIRECTIONAL_STREAMED body mode.
+  // Only one of ``streamed_resp`` or above ``mutation`` should be set.
+  StreamedBodyResponse streamed_resp = 3;
```

Comment:
I'm missing why this is not part of the `oneof` above if it is mutually exclusive with it. Can you provide some info here?

Reply:
Will it cause compatibility issues if adding a new field into an existing oneof?

Reply:
The use of prioritization over oneof was suggested in the xDS-API to allow clients and servers that "speak" using a different version of the protocol (i.e., in a case where the server is aware of the new field but the client isn't, the server will send both the old field and the new field - clients that are not aware of the new field will use the old one, and clients that are aware will use the new one).
AFAIU the current change is related to the ext-proc protocol. If the same logic as above should be applied, then the code and comments should clarify which field will take prioritization if both are included. If not, then the `oneof` should be used.

Reply:
I prefer leaving the new field as a new field instead of putting it inside the oneof. Here are the reasons:
1) If putting it inside the oneof, in case the server is running the new version, and Envoy is running old version, then it will cause Envoy errors if server send a message with the new streamed_resp field encoded.
2) By keeping "streamed_resp" as a separate new field:
   if server runs the new version, envoy runs old version, and if servers sends the a message with the new streamed_resp field encoded, Envoy will just ignore it. The end behavior will be same as server response does not include any change.
    if server runs the old version, envoy runs new version, then server will only be able to send old fields, which is supported by new Envoy.

Reply:
> I prefer leaving the new field as a new field instead of putting it inside the oneof. Here are the reasons:
> 
> 1. If putting it inside the oneof, in case the server is running the new version, and Envoy is running old version, then it will cause Envoy errors if server send a message with the new streamed_resp field encoded.

I'm assuming that server is the side-stream.
In the case where the Envoy has an old proto version and the side-stream has a new proto version, can you explain why would a side-stream sending a message with only streamed_resp field cause an error (whether it is both in the oneof or not)?

> 2. By keeping "streamed_resp" as a separate new field:
>    if server runs the new version, envoy runs old version, and if servers sends the a message with the new streamed_resp field encoded, Envoy will just ignore it. The end behavior will be same as server response does not include any change.
>    if server runs the old version, envoy runs new version, then server will only be able to send old fields, which is supported by new Envoy.

I'm not sure I fully grok why is the difference between an Envoy that has the old-version proto and a side-stream that has a new-version proto.
The only benefit I see for having the field out of the `oneof` is to allow the side-stream to send a single message with both "old" setting, and a "new" (streamed_resp) setting. If that is intentional, please clarify it in the comments (giving precedence to some field over another), and ensure the code handles that.

FWIW, I'm not fully convinced that supporting a gradual rollout is really be needed in the ext_proc case, but I leave it up to you to decide.

Reply:
One message with both old and new setting is not allowed. Also, it's true that if the server is updated to a new version which send the new field, it's reasonable to require Envoy had already been updated to the new version which can handle the new field.
 Moved the new field into oneof.

Reply:
As mentioned above, moved the new field into oneof. 

Any further comments on the API?